### PR TITLE
ARROW-9344: [C++][Flight] Measure latency quantiles

### PR DESCRIPTION
**Pull request by cyb70289**
Measure average, median, 95%, 99% quantile and maximal latency.

### CODE REVIEW

**Comment by cyb70289** [comment]
Please **NOTE** this patch uses `boost::accumulators` package, have to install libboost-dev package before building.

**Comment by lidavidm** [comment]
> Please **NOTE** this patch uses `boost::accumulators` package, have to install libboost-dev package before building.

Hmm, the build system should take care of getting boost for you (unless you intentionally are building with system dependencies, which is OK).

File: cpp/src/arrow/flight/flight_benchmark.cc [hunk 1]
```diff
@@ -281,9 +313,12 @@ Status RunPerformanceTest(FlightClient* client, bool test_put) {
   // Calculate throughput(IOPS) and latency vs batch size
   std::cout << "Throughput: " << (static_cast<double>(stats.total_batches) / time_elapsed)
             << " batches/s" << std::endl;
-  std::cout << "Latency: "
-            << (static_cast<double>(stats.total_nanos) / 1000 / stats.total_batches)
-            << " usec/batch" << std::endl;
+  std::cout << "Latency mean: " << stats.mean_latency() << " uses" << std::endl;
```
Comment (lidavidm) [inline cpp/src/arrow/flight/flight_benchmark.cc:316]: nit: `uses` -> `usec` (or just `us`) and the same below

File: cpp/src/arrow/flight/flight_benchmark.cc [hunk 2]
```diff
@@ -62,20 +69,40 @@ struct PerformanceResult {
...
     std::lock_guard<std::mutex> lock(this->mutex);
     this->total_batches += total_batches;
     this->total_records += total_records;
     this->total_bytes += total_bytes;
-    this->total_nanos += total_nanos;
+  }
+
+  // Invoked per batch in the test loop. Holding a lock looks not scalable.
+  // Tested with 1 ~ 8 threads, no noticeable overhead is observed.
+  // A better approach may be leveraging a lockless queue.
```
Comment (lidavidm) [inline cpp/src/arrow/flight/flight_benchmark.cc:93]: If this is a concern, we could make this object per-thread and merge at the end.

**Review by lidavidm — COMMENTED** [review]
Looks good overall (one minor nit). Do you have sample results?

**Review by IhostVlad — APPROVED** [review]

**Comment by cyb70289** [comment]
> Looks good overall (one minor nit). Do you have sample results?

Test result on my host
```
$ release/arrow-flight-benchmark 
Using standalone server: false
Server running with pid 21435
Testing method: DoGet
Server host: localhost
Server port: 31337
Server host: localhost
Server port: 31337
Batch size: 131040
Batches read: 9768
Bytes read: 1280000000
Nanos: 151282157
Speed: 8069.05 MB/s
Throughput: 64568.1 batches/s
Latency mean: 52 uses
Latency quantile=0.5: 44 usecs
Latency quantile=0.95: 96 usecs
Latency quantile=0.99: 1918 usecs
Latency max: 8513 uses
```

File: cpp/src/arrow/flight/flight_benchmark.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/flight/flight_benchmark.cc:93]: Hmm, merging quantiles looks not straightforward, boost is using [p-square](https://www.cse.wustl.edu/~jain/papers/ftp/psqr.pdf) algorithm which *estimates* the quantile without storing any value.
If there's no much variance among threads, maybe we can simply average per thread quantiles or pick the max. Will do some investigations.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Thank you!

**Pull request closed without merging**